app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionController.scala

```diff
@@ -187,33 +208,81 @@ class SubscriptionController @Inject()(
             sessionStoreService.fetchInitialDetails.flatMap { maybeInitialDetails =>
               maybeInitialDetails
                 .map { initialDetails =>
-                  subscriptionService.subscribe(initialDetails, validDesAddress).map(redirectSubscriptionResponse)
+                  subscriptionService.subscribe(initialDetails, validDesAddress).flatMap(redirectSubscriptionResponse)
                 }
                 .getOrElse(Future.successful(sessionMissingRedirect()))
           }
         )
     }
   }
 
+  val showLinkAccount: Action[AnyContent] = Action.async { implicit request =>
+    withAuthenticatedAgent {
+      Future.successful {
+        request.session.get("arn").fold(sessionMissingRedirect()) { _ =>
+          Ok(html.link_account(linkAccountForm))
+        }
+      }
+    }
+  }
+
+  val submitLinkAccount: Action[AnyContent] = Action.async { implicit request =>
+    withAuthenticatedAgent {
+      request.session.get("arn").fold(Future.successful(sessionMissingRedirect())) { arn =>
+        linkAccountForm
+          .bindFromRequest()
+          .fold(
+            formWithErrors => {
+              if (formWithErrors.errors.exists(_.message == "error.link-account-value.invalid")) {
+                throw new BadRequestException("Form submitted with strange input value")
+              } else {
+                Future.successful(Ok(html.link_account(formWithErrors)))
+              }
+            },
+            validatedLinkAccount => {
+              validatedLinkAccount.autoMapping match {
+                case Yes => linkAccountResponse(mappingConnector.updatePreSubscriptionWithArn)
+                case No  => linkAccountResponse(mappingConnector.deletePreSubscription)
+              }
+            }
+          )
+      }
+    }
+  }
+
   val showSubscriptionComplete: Action[AnyContent] = Action.async { implicit request =>
+    def recoverSessionStoreWithNone[T]: PartialFunction[Throwable, Option[T]] = {
+      case NonFatal(ex) =>
+        Logger(getClass).warn("Session store service failure", ex)
+        None
+    }
+
     withAuthenticatedAgent {
-      request.flash.get("arn") match {
+      request.session.get("arn") match {
```

Comment:
Looks like calling sessionMissingRedirect is repeated scenario when request.session.get("arn") is empty. We can have a common method for that. Can do it future..

Reply:
For future tech debt/discussion

Reply:
Captured in tech debt ticket ASA-71